### PR TITLE
Consolidate e2e log collection

### LINKED ISSUE #1372: Consolidate log collection scripts for e2e tests
/kind cleanup

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
The upstream tests collect logs in different ways.  

The conformance tests collect logs via the [e2e framework](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/test/e2e/azure_logcollector.go).  An example output is: https://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/capz-conformance-master/1391754619595526144/artifacts/

The provider specific tests collect logs [via a script](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/hack/log/log-dump.sh). An example output is: https://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/capz-azure-file-machinepool-master/1391279987113332736/artifacts/

**What did you expect to happen:**
Logs to be collected in same format for easier debugging and consistent output.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by jsturtevant**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
We have several different entry points for our CI and some of them use different logging scripts.  This creates a binary that re-uses to the same logging as our conformance and e2e.  It takes some basic configuration and runs the same logging functions used by the e2e_testsuite.  An added benefit is that a user could also run this and collect logs that are the same as the e2e test.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1372

**Special notes for your reviewer**:

This required moving around some of the capz functions in the e2e test pacakge similiar to what was done in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2130

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Consolidate e2e logging methods
```

### CODE REVIEW

**Comment by Jont828** [comment]
@jsturtevant Is this ready for review or is it still WIP? Saw that there was a broken link in the tests.

**Comment by jsturtevant** [comment]
The other tests are passing, The folder structure isn't 100% the same so going to look into that but the general approach is ready for review.

The broken link is to a file that is in this PR, so not sure how we handle that.  It doesn't exist but when this gets published it would exist...

**Comment by Jont828** [comment]
You can use `<!-- markdown-link-check-disable-line -->` to have it ignore the broken link for now. We use it for when we reference `http://localhost:<port>` as well.

**Comment by jsturtevant** [comment]
> You can use `<!-- markdown-link-check-disable-line -->` to have it ignore the broken link for now. We use it for when we reference `http://localhost:<port>` as well.

Then we follow up with a separate PR to remove it?  I feel that it would be better to merge knowing the check is only failing a link in that will be there once this merges but let me know what we typically do for this scenario.

**Comment by jsturtevant** [comment]
The folder structure for[ ci-entrypoint.sh](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2345/pull-cluster-api-provider-azure-ci-entrypoint/1532800412661321728/artifacts/) is now the same as the [ci-e2e.sh](https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2345/pull-cluster-api-provider-azure-e2e/1532800412564852736/artifacts/) and ci-conformance.sh:

```
artifacts/
└── clusters
    ├── bootstrap
    │   ├── controllers
    │   │   ├── capi-controller-manager
    │   │   ├── capi-kubeadm-bootstrap-controller-manager
    │   │   ├── capi-kubeadm-control-plane-controller-manager
    │   │   └── capz-controller-manager
    │   └── resources
    │       └── default
    │           ├── AzureCluster
    │           ├── AzureClusterIdentity
    │           ├── AzureMachine
    │           ├── AzureMachineTemplate
    │           ├── Cluster
    │           ├── KubeadmConfig
    │           ├── KubeadmConfigTemplate
    │           ├── KubeadmControlPlane
    │           ├── Machine
    │           ├── MachineDeployment
    │           ├── MachineHealthCheck
    │           └── MachineSet
    └── capz-conf-r00mm3
        └── azure-activity-logs
        └── kube-system
        └── machines
            ├── capz-conf-r00mm3-md-win-8675ffb97b-8z89g 
            └── capz-conf-r00mm3-md-win-8675ffb97b-xmmhl

```

**Comment by jsturtevant** [comment]
fyi @lzhecheng @nilo19 @andyzhangx

**Comment by CecileRobertMichon** [comment]
> The broken link is to a file that is in this PR, so not sure how we handle that. It doesn't exist but when this gets published it would exist...

this is a chicken-and-egg problem... no matter what it will require a second PR if we want to job to pass on this PR. We can also ignore the linter and merge the PR with the failure but that's not ideal.

**Comment by jsturtevant** [comment]
> this is a chicken-and-egg problem... no matter what it will require a second PR if we want to job to pass on this PR. We can also ignore the linter and merge the PR with the failure but that's not ideal.

Which approach is preferred?

**Comment by Jont828** [comment]
An alternative approach would be to open a stub PR including the just the file we are linking in the docs (assuming the file isn't going to produce compile errors without the rest being present). Then, you could rebase this PR off of that one. Otherwise, I would personally lean towards using the disable link check for now.

File: test/logger.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+func Fail(message string, callerSkip ...int) {
+	panic(message)
+}
+
+func main() {
+	// needed for ginkgo/gomega which is used by the capi test framework
+	RegisterFailHandler(Fail)
+
+	// required flags
+	kubeconfigPath := flag.String("kube-config", getKubeConfigPath(), "The kubeconfig for the management cluster")
```
Comment (CecileRobertMichon) [inline test/logger.go:45]: nit: might be good to match flag names we already have in other tools like clusterctl and capz for consistency, eg:

`--name` for cluster name, `kubeconfig` for kubeconfig, `--resource-group -g` for resource group

Comment (jsturtevant) [reply test/logger.go:45]: this makes sense, will do

File: test/logger.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	clustername := flag.String("clustername", "", "workload cluster to collect logs for")
+
+	// optional flags that default
+	namespace := flag.String("namespace", "", "namespace on management cluster to collect logs for")
+	artifactFolder := flag.String("artifacts-folder", getArtifactsFolder(), "folder to store cluster logs")
+	azureresourcegroup := flag.String("azure-rg", "", "azure resource group that contains the cluster")
```
[the remark this answers is not in the document]
Comment (jsturtevant) [reply test/logger.go:51]: I hadn't considered using the cluster spec, I will see if that can simplify adding an additional flag

File: test/logger.go (hunk 1 above)
Comment (jsturtevant) [reply test/logger.go]: done

File: test/logger.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsturtevant) [reply test/logger.go]: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2345/files#diff-34d5acd96ae3a8b3040a0d8a711797ef5eaceabe05de690d663fa6cb670cead5R221-R227

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Review by CecileRobertMichon — COMMENTED** [review]
/assign @mboersma

**Review by mboersma — APPROVED** [review]
/lgtm

This is a great improvement! Thanks for addressing my ticky-tack comments.

**Comment by CecileRobertMichon** [comment]
```
--- FAIL: TestAzureMachinePool_Validate (0.00s)
    --- FAIL: TestAzureMachinePool_Validate/HasValidImage (0.00s)
        azuremachinepool_test.go:66: 
            Unexpected error:
                <*field.Error | 0xc00031c040>: {
                    Type: "FieldValueForbidden",
                    Field: "spec",
                    BadValue: <string>"",
                    Detail: "can be set only if the MachinePool feature flag is enabled",
                }
                spec: Forbidden: can be set only if the MachinePool feature flag is enabled
            occurred
 ```

This failure seems to be coming from https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2376 which is strange because it passed all tests before merging

**Comment by jsturtevant** [comment]
is it happening consistently and for other prs?

/test pull-cluster-api-provider-azure-test

**Comment by CecileRobertMichon** [comment]
hmm retest passed 🤔 

I have not seen it on other PRs, but let's keep an 👁️  out. If it's non-deterministic that's an issue.

**Pull request merged**